packages/jest-config/package.json

```diff
@@ -27,7 +27,8 @@
     "jest-util": "^26.3.0",
     "jest-validate": "^26.4.2",
     "micromatch": "^4.0.2",
-    "pretty-format": "^26.4.2"
+    "pretty-format": "^26.4.2",
+    "typescript": "^4.0.3"
```

Comment:
This is a really big dependency (~40mb iirc at least, if not double), in addition to pinning the ts version.

I think the better way to support ts would be to use `ts-node`, ideally as an optional dependency of some sort in the same way webpack does (or alternatively doing that with `typescript`).

Alternatively you could mabe use Babel?

Jest already has decent config validation that'll result in messages that are a lot nicer than TypeScripts, as TS will just say "this type is wrong", so (imo) supporting `jest.config.ts` should be about supporting taking in a config that's written in TS without typechecking, meaning we can use a more lightweight way of doing that.

I should have some notes from my `.eslintrc.ts` RFC that describes how Webpack & co did it - I'll dig that up for you, as it should be a lot more lightweight.

Reply:
(in saying this, if the jest team are happy for `typescript` to be a dependency, then going with that is fine?)

Reply:
That's right...

so `webpack-nano` uses [`rechoir`](https://github.com/gulpjs/rechoir), and [`karma`](https://github.com/karma-runner/karma/blob/7617279f68507642fdc7eb0083916f3c6c2a28cb/lib/config.js#L36-L39) does this:

```
try {
  require('ts-node').register()
  TYPE_SCRIPT_AVAILABLE = true
} catch (e) {}
```

Personally I think something close to the `karma` way would be the best solution, as it's lightweight - the idea is basically "get out of ts asap in the least number of steps".

This should be fine dependency-wise as `ts-node` can be listed as an optional peerDependency.

Reply:
@G-Rath Sounds good, I will look into this.

Reply:
tbh the patch I wrote for implementing support in `eslint` is pretty much exactly what you want:

```
/**
 * Loads a TypeScript configuration from a file.
 * @param {string} filePath The filename to load.
 * @returns {ConfigData} The configuration object from the file.
 * @throws {Error} If the file cannot be read.
 * @private
 */
function loadTSConfigFile(filePath) {
  debug(`Loading TS config file: ${filePath}`);

  let registerer;

  try {
    registerer = require('ts-node').register();
  } catch (e) {
    debug(`Could not register ts-node to read TypeScript file: ${filePath}`);
    e.message = `ts-node is required for TypeScript configuration files. Make sure it is installed\nError: ${e.message}`;
    throw e;
  }

  try {
    registerer.enabled(true);
    const imported = importFresh(filePath);
    registerer.enabled(false);

    return imported;
  } catch (e) {
    debug(`Error reading TypeScript file: ${filePath}`);
    e.message = `Cannot read config file: ${filePath}\nError: ${e.message}`;
    throw e;
  }
}
```

That handles requiring `ts-node` with the register, requiring the config (just replace `importFresh` with `require`), and the disabling the ts-node registerer so it doesn't mess with future requires.

Disabling the registerer is super important for `jest` because it has its own transformations that'll get messed up if `ts-node` transpiles test files first.